core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/LmdbSailStore.java

```diff
@@ -561,14 +571,28 @@ private void startTransaction(boolean preferThreading) throws SailException {
 														op.execute();
 													}
 												} else {
-													Thread.yield();
+													if (!running.get()) {
+														logger.warn(
+																"LmdbSailStore was closed while active transaction was waiting for the next operation. Forcing a rollback!");
+														opQueue.add(ROLLBACK_TRANSACTION);
+													} else if (Thread.interrupted()) {
+														throw new InterruptedException();
+													} else {
+														Thread.yield();
+													}
```

Comment:
@kenwenzel I've made some tests to fix some deadlock scenarios in the MemoryStore. The tests open two connections and start two transactions, in two separate threads. One of the tests then tries to call shutdown() on the store, the other test tries to close the connections directly. I ran into an issue with the async transaction handling in the LmdbSailStore that I've tried to handle here.

The issue is that when the executor service is shutdown() it waits indefinitely for the async transaction handling to complete. The async transaction handling can at that point be stuck in a `while(true)` loop that is waiting for the transaction to either be committed or rolled back. For some reason it doesn't seem to get rolled back, but adding a check for `running` and then rolling back the transaction seems to work.

What do you think @kenwenzel ?

Reply:
Thank you for Investigating this. I am not quite sure why this happens as I thought that it is handled by the running flag. But maybe I was wrong.